contributors/design-proposals/local-pv-provisioner.md

```diff
@@ -0,0 +1,185 @@
+
+### Master Provisioner
+The populator is initialized so that it calls the appropriate gRPC for every PV Informer update. Then it populates each worker's cache by calling `PushCurrentPVs()`, passing the collection of PVs to be controlled by this worker.
+
+The master then performs worker discovery as described in the [Worker Discovery](#worker-discovery) section of design 1, and spawns a goroutine for each worker. Each goroutine periodically calls `GetPVRequest()`, performing that request if it receives one. Then it calls `CompletePVRequest()` to signal the request completion.
+
+## Design 3: Event-Driven with Custom Resources
+
+### High Level Overview
+The previous two designs are complicated by the asynchronous nature of delete and the presence of failures. 
+As recommended by @dhirajh, another design is to persist ProvisionerStatus as a Custom Resource, and to drive all operations with PV and ProvisionerStatus Informer events.
+
+To ensure no malicious pods can affect provisioner execution, the CustomResourceDefinition and all pods must be in a trusted admin namespace.
+
+In the following description, we assume all Informer updates are cached and events buffered. On initialization, all caches are populated, and every Informer update event handler is executed once to ensure the entire provisioner system is in a consistent state.
+
+### ProvisionerStatus CRD
+The CRD contains the following fields (similar to design 1):
+
+* `LocalVolumes` - A collection of volumes mounted in the local storage directory on the node. The discoverer is triggered to acquire this information.
+* `DeleteOperations` - A list keeping track of information related to each delete operation. This status is managed by the cleaner. It contains the following fields:
+  * PV - The PV to be deleted on disk.
+  * DeleteStatus - takes one of the following values:
+    * Requested
+    * InProgress
+    * Complete
+    * Error
+    
+There exists one ProvisionerStatus object per node.
+
+To ensure correctness, master and worker never write to the same field. Specifically:
+* After a `DeleteOperation` is created, its `PV` field is never modified.
+* Master can only create and delete `DeleteOperation`s. It never modifies those objects.
+* Worker can only modify `DeleteStatus`. It can never create or delete `DeleteOperation`s.
+* Only a worker can write to `LocalVolumes`.
+
+### Worker Provisioner
+Create the ProvisionerStatus object on initialization.
+
+The Discoverer periodically probes the filesystem and updates ProvisionerStatus `LocalVolumes`. This can be further improved by using a filesystem watch.
+
+The Deleter uses ProvisionerStatus to keep track of states for each PV being deleted. Deletion is started when the ProvisionerStatus is updated (by the master) and some PVs are in the Requested state. Details will be included in the async volume cleaner design.
```

Comment:
Sorry, `some PVs are in the Requested state` means: `DeleteStatus` value is `Requested`, right ?

Reply:
Yes